src/metrics.rs

```diff
@@ -455,26 +464,81 @@ impl GooseRequestMetricAggregate {
-                self.requests_per_second.push(0);
-            }
-        }
-
+        expand_per_second_metric_array(&mut self.requests_per_second, second, 0);
         self.requests_per_second[second] += 1;
 
         debug!(
-            "incremented second {} counter: {}",
+            "incremented second {} for requests per second counter: {}",
             second, self.requests_per_second[second]
         );
     }
+
+    /// Record errors per second metric.
+    pub(crate) fn record_errors_per_second(&mut self, second: usize) {
+        expand_per_second_metric_array(&mut self.errors_per_second, second, 0);
+        self.errors_per_second[second] += 1;
+
+        debug!(
+            "incremented second {} for errors per second counter: {}",
+            second, self.errors_per_second[second]
+        );
+    }
+
+    /// Record average response time per second metric.
+    pub(crate) fn record_average_response_time_per_second(
+        &mut self,
+        second: usize,
+        response_time: u64,
+    ) {
+        expand_per_second_metric_array(
+            &mut self.average_response_time_per_second,
+            second,
+            (0, 0.0),
+        );
+        // First element of the tuple is a number of data points that we used
+        // to calculate the average so far and the second element of the tuple
+        // is the current average value.
+        self.average_response_time_per_second[second].0 += 1;
+        self.average_response_time_per_second[second].1 += (response_time as f32
+            - self.average_response_time_per_second[second].1)
+            / self.average_response_time_per_second[second].0 as f32;
+
+        debug!(
+            "updated second {} for average response time per second: {}",
+            second, self.average_response_time_per_second[second].1
+        );
+    }
+}
+
+/// Expands vectors that collect per-second data for HTML report graphs with a
+/// default value.
+///
+/// We need to do that since we don't know for how long the load test will run
+/// and we can't initialize these vectors at the beginning. It is also
+/// better to do it as we go to save memory.
+fn expand_per_second_metric_array<T: Clone>(data: &mut Vec<T>, second: usize, initial: T) {
+    // Each element in per second metric vectors (self.requests_per_second,
+    // self.errors_per_second, ...) is count for a given second since the start
+    // of the test. Since we don't know how long the test will at the beginning
+    // we need to push new elements (second counters) as the test is running.
+    if data.len() <= second {
+        for _ in 0..(second - data.len() + 1) {
+            data.push(initial.clone());
+        }
+    };
 }
+
+/// Implement equality for GooseRequestMetricAggregate. We can't simply derive since
+/// we have floats in the struct.
+///
+/// `Eq` trait has no functions on it - it is there only to inform the compiler
+/// that this is and equivalence rather than a partial equivalence.
```

Comment:
typo: `this is an equivalence` not `this is and...`

Reply:
Done.